### PR TITLE
Bulk upsert in DocumentFetcher with duplicate detection

### PR DESCRIPTION
**Pull request by yoomlam**
Addresses problems when deploying #15707 on real documents.

### Description
Sends warnings about duplicates and removes duplicates so that bulk upsert doesn't fail.

### Acceptance Criteria
- [ ] Code compiles correctly

### Testing Plan
1. Deploy during low-user activity
2. Run Reader tests
3. Check Sentry for warnings to diagnose cause for duplication

### CODE REVIEW

File: app/models/document.rb [hunk 1]
```diff
@@ -215,19 +248,19 @@ def content_url
...
 
   def copy_metadata_from_document(source_document)
-    source_document.annotations.map do |annotation|
+    new_annotations = source_document.annotations.map do |annotation|
       annotation.dup.tap do |a|
         a.document_id = id
-        a.save!
       end
     end
+    Annotation.create(new_annotations.map(&:attributes))
```
Comment (yoomlam) [inline app/models/document.rb:256]: Create all annotations using one SQL query.

File: app/models/document.rb [hunk 2]
```diff
@@ -215,19 +248,19 @@ def content_url
...
 
-    source_document.documents_tags.map do |tag|
+    new_tags = source_document.documents_tags.map do |tag|
       tag.dup.tap do |t|
         t.document_id = id
-        t.save!
       end
     end
+    DocumentsTag.create(new_tags.map(&:attributes))
```
Comment (yoomlam) [inline app/models/document.rb:263]: Create all tags using one SQL query.

File: app/services/document_fetcher.rb [hunk 3]
```diff
@@ -37,40 +38,59 @@ def fetched_at_format
...
-      accumulator[document.vbms_document_id] = document
-    end
+    # Each document has a series_id and a version_id
+    # (unfortunately we refer to version_id as vbms_document_id in most of the code)
+    # See https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Reader#vbms
+    vbms_doc_ver_ids = documents.map(&:vbms_document_id)
 
-    documents.map do |document|
-      if existing_documents[document.vbms_document_id]
-        document.merge_into(existing_documents[document.vbms_document_id]).save!
```
Comment (yoomlam) [inline app/services/document_fetcher.rb:50]: `save!` is being called for each doc! This PR replaces this with a bulk save using `Document.bulk_merge_and_save` and `Document.import`.

File: spec/services/document_fetcher_spec.rb [hunk 4]
```diff
@@ -314,8 +316,38 @@
           # Uncomment the following to see a count of SQL queries
           # pp query_data.values.pluck(:sql, :count)
           doc_insert_queries = query_data.values.select { |o| o[:sql].start_with?("INSERT INTO \"documents\"") }
-          # This expected count of 25 is bad and will be fixed in another PR
-          expect(doc_insert_queries.pluck(:count).max).to eq 25
```
Comment (yoomlam) [inline spec/services/document_fetcher_spec.rb:318]: Now fixed in this PR!

File: app/services/document_fetcher.rb [hunk 5]
```diff
@@ -37,40 +38,59 @@ def fetched_at_format
...
-      end
-    rescue ActiveRecord::RecordNotUnique
-      Document.find_by_vbms_document_id(document.vbms_document_id)
-    end
-  end
+    existing_vbms_doc_ver_ids = Document.where(vbms_document_id: vbms_doc_ver_ids).pluck(:vbms_document_id)
+    docs_to_update, docs_to_create = split_docs(documents, existing_vbms_doc_ver_ids)
+
+    # Update existing docs
+    docs_to_update = deduplicate(docs_to_update, "docs_to_update") # To-do: remove after seeing no duplicate warnings
```
Comment (yoomlam) [inline app/services/document_fetcher.rb:53]: This line *should* not be needed, but inserting it for debugging.

File: app/services/document_fetcher.rb [hunk 6]
```diff
@@ -37,40 +38,59 @@ def fetched_at_format
...
+    updated_docs = Document.bulk_merge_and_save(docs_to_update)
 
-  def create_new_document!(document, ids)
-    document.save!
+    # Create new docs that don't exist
+    docs_to_create = deduplicate(docs_to_create, "docs_to_create") # To-do: remove after seeing no duplicate warnings
```
Comment (yoomlam) [inline app/services/document_fetcher.rb:57]: This line *should* not be needed, but inserting it for debugging.

File: spec/services/document_fetcher_spec.rb [hunk 7]
```diff
@@ -314,8 +316,38 @@
...
+          expect(doc_insert_queries.pluck(:count).max).to eq 1
+          expect(query_data.values.select { |o| o[:sql].start_with?("UPDATE") }).to be_empty
+        end
+
+        context "when there are duplicate documents returned from document_service" do
```
Comment (yoomlam) [inline spec/services/document_fetcher_spec.rb:323]: Was not able to replicate this scenario in prod. Create the scenario here and send enough data to Sentry for analysis against real documents in prod.

File: spec/services/document_fetcher_spec.rb [hunk 8]
```diff
@@ -111,7 +112,7 @@
     context "when there is no existing document" do
       it "saves retrieved documents" do
         returned_documents = document_fetcher.find_or_create_documents!
-        expect(returned_documents.map(&:type)).to eq(documents.map(&:type))
+        expect(returned_documents.map(&:type)).to match_array(documents.map(&:type))
```
Comment (yoomlam) [inline spec/services/document_fetcher_spec.rb:115]: Reduce flakes.

**Review by lomky — APPROVED** [review]
Looks good!

**Comment by yoomlam** [comment]
[Sentry alerts](https://sentry.prod.appeals.va.gov/department-of-veterans-affairs/caseflow/issues/13513/events/) for investigation.
  * [352 distinct docs duplicated](https://sentry.prod.appeals.va.gov/department-of-veterans-affairs/caseflow/issues/13513/events/992116/)
  * [46 out of ~588 docs duplicated](https://sentry.prod.appeals.va.gov/department-of-veterans-affairs/caseflow/issues/13513/events/992115/)

[New Relic](https://dsva.slack.com/archives/CHX8FMP28/p1610753072027900?thread_ts=1610749693.025200&cid=CHX8FMP28) looks promising.

Conclusions so far:
* docs with the same `vbms_document_id` are indeed duplicates (other attributes have the same values)
* dedupping in `fetch_documents_from_service!` is sufficient; dedupping on lines 53 and 57 are not needed.
* TODO: Investigate why the document_service is returning duplicate records

**Pull request merged**